src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -9,16 +9,26 @@
 #include "mpir_nodemap.h"
 #include "mpidu_init_shm.h"
 #include "xpmem_seg.h"
+#include <sys/stat.h>
 
 int MPIDI_XPMEM_mpi_init_hook(int rank, int size, int *n_vcis_provided, int *tag_bits)
 {
     int mpi_errno = MPI_SUCCESS;
-    int i;
+    struct stat buf;
 
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_INIT_HOOK);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_INIT_HOOK);
     MPIR_CHKPMEM_DECL(1);
 
+    /* While this version of MPICH does require libxpmem to link, we don't necessarily require the
+     * kernel module to be loaded at runtime. If XPMEM is not available, disable its use via the
+     * special CVAR value. */
+    if (0 != stat("/dev/xpmem", &buf)) {
+        MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE = -1;
+        MPL_DBG_MSG_FMT(MPIR_DBG_PT2PT, VERBOSE, (MPL_DBG_FDEST,
+                                                  "/dev/xpmem not found. Disabling XPMEM support"));
+        goto fn_exit;
+    }
 #ifdef MPL_USE_DBG_LOGGING
```

Comment:
Yeah, that's fair. My specific use case was that I know the XPMEM library will be available, but the kernel module might not be loaded. When the kernel module is loaded `/dev/xpmem` appears. If that's not true for you (or the use case you're targeting) this can change.

Reply:
Maybe we should let configure check kernel module too. E.g., execute a program with xpmem sys call.

Reply:
`xpmem_init` will fail if it can't find the `/dev/` file. https://github.com/hjelmn/xpmem/blob/master/lib/libxpmem.c#L32. Do we ignore that failure, or do we crash and hint the user either to reconfigure, or set an envvar to disable?

Reply:
@raffenet You are right. The master code should report an error when calling `xpmem_make` at `MPIDI_XPMEM_mpi_init_hook` (`xpmem_make` internally calls `xpmem_init`). This is a runtime failure. We can improve the error message to ask user to load the kernel module, or disable xpmem at runtime (with this patch), or disable xpmem at configure time (with master)

The benefit of this commit is to disable xpmem only on the machines that do not load the xpmem module. I do not think this is a common case for most MPI users. If we want to have this feature, maybe we can just handle the `xpmem_make` failure in xpmem_init rather than throwing out the error.

Reply:
> Maybe we should let configure check kernel module too. E.g., execute a program with xpmem sys call.

I'm not sure you'd want to do this. You might want to build somewhere that the user library is install, but the module isn't loaded.

Reply:
> The benefit of this commit is to disable xpmem only on the machines that do not load the xpmem module. I do not think this is a common case for most MPI users. If we want to have this feature, maybe we can just handle the `xpmem_make` failure in `MPIDI_XPMEM_mpi_init_hook ` rather than throwing out the error.

I agree. It would be nice to be able to just silently switch over (or print something or put something in the logs...).

Reply:
@wesbland Do you plan to replace the commit with `xpmem_make` error handling in this PR ?

Reply:
Sure. I can do that pretty quickly.

Reply:
Done.